wrappers/python/nntile/layer/llama_attention.py

```diff
@@ -0,0 +1,1330 @@
+        if bias:
+            out_proj_bias_traits = TensorTraits([n_emb], [n_emb_tile])
+            out_proj_bias_distr = [0] * out_proj_bias_traits.grid.nelems
+            out_proj_bias_value = type(x.value)(
+                out_proj_bias_traits, out_proj_bias_distr, next_tag
+            )
+            next_tag = out_proj_bias_value.next_tag
+            out_proj_bias_grad = type(x.value)(
+                out_proj_bias_traits, out_proj_bias_distr, next_tag
+            )
+            next_tag = out_proj_bias_grad.next_tag
+            out_proj_bias = TensorMoments(
+                out_proj_bias_value, out_proj_bias_grad, True
+            )
+        else:
+            out_proj_bias = None
+        # Allocate tensor for output y
+        y_traits = TensorTraits(x.value.shape, x.value.basetile_shape)
+        y_value = type(x.value)(y_traits, x.value.distribution, next_tag)
+        next_tag = y_value.next_tag
+        y_grad = type(x.value)(y_traits, x.value.distribution, next_tag)
+        next_tag = y_grad.next_tag
+        y = TensorMoments(y_value, y_grad, True)
+        # Create attention layer with all the provided data
+        layer = LlamaAttention(
+            x,
+            y,
+            w_q,
+            w_k,
+            w_v,
+            w,
+            q_transposed,
+            q,
+            k_transposed,
+            k,
+            k_rep,
+            v_transposed,
+            v,
+            v_rep,
+            a,
+            a_maxsumexp,
+            a_sumprod_slice,
+            b,
+            b_transposed,
+            bias_inproj_q,
+            bias_inproj_k,
+            bias_inproj_v,
+            out_proj_bias,
+            mask,
+            redux=redux,
+        )
+        # Return layer and next tag to be used
+        return (layer, next_tag)
+
+    # Forward propagation of the attention layer
+    def forward_async(self):
+        # Compute query, key and value tensors
+        # Q_transposed = einsum('ijkl,lmn->ijkmn', W_Q, X_Q)
+        # gemm (kv_group_size, n_head_kv, head_size, n_emb)
+        # by (n_emb, n_seq, n_batch)
+        # into (kv_group_size, n_head_kv, head_size, n_seq, n_batch)
+        gemm_async(
+            1.0,
+            notrans,
+            self.w_q.value,
+            notrans,
+            self.x_q.value,
+            0.0,
+            self.q_transposed.value,
+            1,
+            0,
+            redux=self.redux,
+        )
+        # Rotate axes into
+        # (head_size, n_seq, n_batch, kv_group_size, n_head_kv)
+        transpose_async(1.0, self.q_transposed.value, self.q.value, 2)
+        # X_Q, W_Q and Q_transposed can be offloaded from GPU
+        self.x_q.value.wont_use()
+        self.q_transposed.value.invalidate_submit()
+        self.w_q.value.wont_use()
+        # Apply bias if needed
+        if self.in_proj_bias_q is not None:
+            # batched add_fiber (head_size, batch=(kv_group_size, n_head_kv))
```

Comment:
What does internal tuple batch=() mean? What shape should return in this case?

Reply:
This is just a virtual union of several dimensions into a single batch dimension. Parameter `batch_ndim=2` is set in the next line.